dbms/src/Interpreters/ActionsVisitor.cpp

```diff
@@ -107,37 +93,77 @@ SetPtr makeExplicitSet(
         return 0;
     };
 
-    size_t left_tuple_depth = getTupleDepth(left_arg_type);
-    size_t right_tuple_depth = getTupleDepth(right_arg_type);
+    const auto& right_arg_type = right_arg_evaluated.second;
+    const auto& right_arg_value = right_arg_evaluated.first;
 
-    ASTPtr elements_ast = nullptr;
-
-    /// 1 in 1; (1, 2) in (1, 2); identity(tuple(tuple(tuple(1)))) in tuple(tuple(tuple(1))); etc.
-    if (left_tuple_depth == right_tuple_depth)
+    const size_t left_tuple_depth = getTupleDepth(left_arg_type);
+    const size_t right_tuple_depth = getTupleDepth(right_arg_type);
+    if (left_tuple_depth != right_tuple_depth && left_tuple_depth + 1 != right_tuple_depth)
     {
-        ASTPtr exp_list = std::make_shared<ASTExpressionList>();
-        exp_list->children.push_back(right_arg);
-        elements_ast = exp_list;
+        throw Exception("Invalid types for IN function: "
+                        + left_arg_type->getName() + " and " + right_arg_type->getName() + ".",
+                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
     }
-    /// 1 in (1, 2); (1, 2) in ((1, 2), (3, 4)); etc.
-    else if (left_tuple_depth + 1 == right_tuple_depth)
+
+    Block block;
```

Comment:
It's better to extract all the logic with temporary block in it's own function like createFromAST, or even into two separate functions: one for numbers and one for Array, Tuple and Set.